File: server/machines/registry.py
"""Load machine profiles and expose lookup helpers."""
from __future__ import annotations

import json
from difflib import get_close_matches
from pathlib import Path
from typing import Dict, Iterable, List, Tuple

MachineProfile = Dict[str, object]
MachineSummary = Dict[str, object]

ROOT = Path(__file__).resolve().parents[2]
MACHINES_DIR = ROOT / "config" / "machines"

_by_id: Dict[str, MachineProfile] = {}
_by_id_lower: Dict[str, MachineProfile] = {}
_by_alias_lower: Dict[str, MachineProfile] = {}
_fuzzy_keys: List[str] = []
_fuzzy_map: Dict[str, MachineProfile] = {}


def _load_profiles() -> Dict[str, MachineProfile]:
    profiles: Dict[str, MachineProfile] = {}
    for path in MACHINES_DIR.glob("*.json"):
        if path.name.startswith("_"):
            continue
        data: MachineProfile = json.loads(path.read_text())
        identifier = str(data["id"]).lower()
        profiles[identifier] = data
    return profiles


def _build_indexes(profiles: Dict[str, MachineProfile]) -> None:
    global _by_id, _by_id_lower, _by_alias_lower, _fuzzy_keys, _fuzzy_map
    _by_id = {profile["id"]: profile for profile in profiles.values()}
    _by_id_lower = {key.lower(): value for key, value in _by_id.items()}
    _by_alias_lower = {}
    _fuzzy_map = {}
    _fuzzy_keys = []
    for profile in _by_id.values():
        identifier = str(profile["id"]).lower()
        _fuzzy_keys.append(identifier)
        _fuzzy_map[identifier] = profile
        aliases = profile.get("aliases") or []
        if isinstance(aliases, list):
            for alias in aliases:
                alias_key = str(alias).lower()
                _by_alias_lower[alias_key] = profile
                _fuzzy_keys.append(alias_key)
                _fuzzy_map[alias_key] = profile
        brand = str(profile.get("brand", "")).strip()
        model = str(profile.get("model", "")).strip()
        if brand or model:
            combo = f"{brand} {model}".strip().lower()
            if combo:
                _fuzzy_keys.append(combo)
                _fuzzy_map[combo] = profile
            if model:
                _fuzzy_keys.append(model.lower())
                _fuzzy_map[model.lower()] = profile


def reload_registry() -> None:
    """Reload machine metadata from disk."""
    profiles = _load_profiles()
    _build_indexes(profiles)


def all_machine_profiles() -> Iterable[MachineProfile]:
    return _by_id.values()


def machine_summaries() -> List[MachineSummary]:
    summaries: List[MachineSummary] = []
    for profile in _by_id.values():
        summaries.append(
            {
                "id": profile.get("id"),
                "brand": profile.get("brand"),
                "model": profile.get("model"),
                "aliases": profile.get("aliases", []),
            }
        )
    summaries.sort(key=lambda item: (str(item.get("brand") or "").lower(), str(item.get("model") or "")))
    return summaries


def resolve_machine(name_or_id: str) -> MachineProfile:
    if not name_or_id:
        raise KeyError("Machine identifier cannot be empty")
    token = name_or_id.strip().lower()
    direct = _by_id_lower.get(token)
    if direct:
        return direct
    alias = _by_alias_lower.get(token)
    if alias:
        return alias
    matches = get_close_matches(token, _fuzzy_keys, n=1, cutoff=0.6)
    if matches:
        match = matches[0]
        resolved = _fuzzy_map.get(match)
        if resolved:
            return resolved
    raise KeyError(f"Machine '{name_or_id}' was not found in the registry")


# Initial load during module import.
reload_registry()

File: server-node/src/machines/registry.ts
import fs from "node:fs";
import path from "node:path";
import { settings } from "../settings";

export type MachineProfile = Record<string, any>;
export type MachineSummary = Record<string, any>;

const MACHINES_DIR = path.resolve(process.cwd(), settings.data.machinesDir);

let byId = new Map<string, MachineProfile>();
let byIdLower = new Map<string, MachineProfile>();
let byAliasLower = new Map<string, MachineProfile>();
let fuzzyKeys: string[] = [];
let fuzzyMap = new Map<string, MachineProfile>();

function loadProfiles(): Map<string, MachineProfile> {
  const profiles = new Map<string, MachineProfile>();
  if (!fs.existsSync(MACHINES_DIR)) return profiles;
  for (const file of fs.readdirSync(MACHINES_DIR)) {
    if (!file.endsWith(".json") || file.startsWith("_")) continue;
    const full = path.join(MACHINES_DIR, file);
    try {
      const data = JSON.parse(fs.readFileSync(full, "utf-8")) as MachineProfile;
      const identifier = String(data?.id ?? path.basename(file, ".json"));
      profiles.set(identifier.toLowerCase(), data);
    } catch (err) {
      // eslint-disable-next-line no-console
      console.warn(`Failed to load machine profile ${file}:`, err);
    }
  }
  return profiles;
}

function buildIndexes(profiles: Map<string, MachineProfile>) {
  byId = new Map();
  byIdLower = new Map();
  byAliasLower = new Map();
  fuzzyKeys = [];
  fuzzyMap = new Map();

  for (const profile of profiles.values()) {
    const identifier = String(profile.id ?? "");
    if (!identifier) continue;
    byId.set(identifier, profile);
    byIdLower.set(identifier.toLowerCase(), profile);
    const idKey = identifier.toLowerCase();
    fuzzyKeys.push(idKey);
    fuzzyMap.set(idKey, profile);

    const aliases: unknown = profile.aliases;
    if (Array.isArray(aliases)) {
      for (const alias of aliases) {
        const key = String(alias).toLowerCase();
        byAliasLower.set(key, profile);
        fuzzyKeys.push(key);
        fuzzyMap.set(key, profile);
      }
    }

    const brand = String(profile.brand ?? "").trim();
    const model = String(profile.model ?? "").trim();
    if (brand || model) {
      const combo = `${brand} ${model}`.trim().toLowerCase();
      if (combo) {
        fuzzyKeys.push(combo);
        fuzzyMap.set(combo, profile);
      }
      if (model) {
        const modelKey = model.toLowerCase();
        fuzzyKeys.push(modelKey);
        fuzzyMap.set(modelKey, profile);
      }
    }
  }
}

function levenshtein(a: string, b: string): number {
  const m = a.length;
  const n = b.length;
  const dp = Array.from({ length: m + 1 }, () => new Array<number>(n + 1).fill(0));
  for (let i = 0; i <= m; i++) dp[i][0] = i;
  for (let j = 0; j <= n; j++) dp[0][j] = j;
  for (let i = 1; i <= m; i++) {
    for (let j = 1; j <= n; j++) {
      const cost = a[i - 1] === b[j - 1] ? 0 : 1;
      dp[i][j] = Math.min(dp[i - 1][j] + 1, dp[i][j - 1] + 1, dp[i - 1][j - 1] + cost);
    }
  }
  return dp[m][n];
}

function similarity(a: string, b: string): number {
  if (!a && !b) return 1;
  const dist = levenshtein(a, b);
  const maxLen = Math.max(a.length, b.length) || 1;
  return 1 - dist / maxLen;
}

export function reloadRegistry(): void {
  const profiles = loadProfiles();
  buildIndexes(profiles);
}

export function allMachineProfiles(): MachineProfile[] {
  if (!byId.size) reloadRegistry();
  return Array.from(byId.values());
}

export function machineSummaries(): MachineSummary[] {
  if (!byId.size) reloadRegistry();
  const summaries: MachineSummary[] = [];
  for (const profile of byId.values()) {
    summaries.push({
      id: profile.id,
      brand: profile.brand,
      model: profile.model,
      aliases: profile.aliases ?? [],
      type: profile.type,
      capabilities: profile.capabilities ?? [],
      safe_speed_ranges: profile.safe_speed_ranges,
      material_presets: profile.material_presets,
      max_nozzle_temp_c: profile.max_nozzle_temp_c,
      max_bed_temp_c: profile.max_bed_temp_c,
      spindle_rpm_range: profile.spindle_rpm_range,
      max_feed_mm_min: profile.max_feed_mm_min,
      supports: profile.supports,
      notes: profile.notes,
    });
  }
  summaries.sort((a, b) => {
    const brandA = String(a.brand ?? "").toLowerCase();
    const brandB = String(b.brand ?? "").toLowerCase();
    if (brandA === brandB) {
      return String(a.model ?? "").localeCompare(String(b.model ?? ""));
    }
    return brandA.localeCompare(brandB);
  });
  return summaries;
}

export function resolveMachine(nameOrId: string): MachineProfile {
  if (!byId.size) reloadRegistry();
  const token = nameOrId.trim().toLowerCase();
  if (!token) throw new Error("Machine identifier cannot be empty");
  const direct = byIdLower.get(token);
  if (direct) return direct;
  const alias = byAliasLower.get(token);
  if (alias) return alias;

  let best: { key: string; score: number } | null = null;
  for (const key of fuzzyKeys) {
    const score = similarity(token, key);
    if (!best || score > best.score) {
      best = { key, score };
    }
  }
  if (best && best.score >= 0.6) {
    const profile = fuzzyMap.get(best.key);
    if (profile) return profile;
  }
  throw new Error(`Machine '${nameOrId}' was not found in the registry`);
}

reloadRegistry();
